### PR TITLE
Add Elem::which_node_am_i() interface and unit tests

### PR DESCRIPTION
**Pull request by jwpeterson**
This function allows you to determine which (local) node number of Element `e` corresponds to (local) node number `n` on side `s`. This should be more efficient than calling `Elem::build_side()` in situations where you only need e.g. DOF values from the parent element and not the full side information.

### CODE REVIEW

File: include/geom/node_elem.h [hunk 1]
```diff
@@ -106,7 +106,7 @@ class NodeElem : public Elem
    * This should never be important for NodeElems.
    */
   virtual dof_id_type key (const unsigned int) const libmesh_override
-  { return 0; }
+  { libmesh_not_implemented(); return 0; }
```
Comment (roystgnr) [inline include/geom/node_elem.h]: This should be libmesh_error(), not libmesh_not_implemented(), right?  libmesh_not_implemented() is for corner cases where the call is sensible but we just haven't written the code, whereas I suspect calling key() on a NodeElem will always be an error..

File: tests/geom/which_node_am_i_test.C [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+  void testPrisms()
+  {
+    // A PRISM6 has four nodes on some sides and three nodes on others
+    const Elem & prism6 = ReferenceElem::get(PRISM6);
+    CPPUNIT_ASSERT_EQUAL(static_cast<unsigned int>(4), prism6.which_node_am_i(/*side=*/4, /*node=*/1));
+
+// #if !defined(NDEBUG) && defined(LIBMESH_ENABLE_EXCEPTIONS)
+//     try
+//       {
```
Comment (roystgnr) [inline tests/geom/which_node_am_i_test.C]: Would it make sense to temporarily swap or otherwise disable libMesh::err before the try block here?  It won't matter if libmesh_assert_foo() is spewing scary-looking messages so long as the messages are being spewed to /dev/null.

File: tests/Makefile.in [hunk 3]
```diff
@@ -187,7 +187,8 @@ am__unit_tests_dbg_SOURCES_DIST = driver.C test_comm.h \
 	fe/fe_l2_lagrange_test.C fe/fe_lagrange_test.C \
 	fe/fe_monomial_test.C fe/fe_szabab_test.C fe/fe_test.h \
 	fe/fe_xyz_test.C geom/elem_test.C geom/node_test.C \
-	geom/point_test.C geom/point_test.h mesh/all_tri.C \
+	geom/point_test.C geom/point_test.h \
+	geom/which_node_am_i_test.C mesh/all_tri.C \
 	mesh/boundary_mesh.C mesh/boundary_info.C mesh/checkpoint.C \
```
Comment (roystgnr) [inline tests/Makefile.in]: Don't we usually edit the build system and re-bootstrap in separate commits?  Slacker.

Comment (roystgnr) [inline tests/Makefile.in]: In fact, I could swear that the reason I always do that is because *you* always catch me when I don't...

Comment (jwpeterson) [inline tests/Makefile.in]: Ugh, yeah. I figured it was OK in this instance since it was only a 1 line change in Makefile.am, but I can split them up.

File: tests/geom/which_node_am_i_test.C (hunk 2 above)
Comment (jwpeterson) [inline tests/geom/which_node_am_i_test.C]: I thought about trying that, but didn't want to mess with it in the current PR. I can do so if you think it's important.

File: include/geom/node_elem.h (hunk 1 above)
Comment (jwpeterson) [inline include/geom/node_elem.h]: Yeah, I'll fix that...

File: src/geom/face_tri6.C [hunk 4]
```diff
@@ -174,6 +174,17 @@ dof_id_type Tri6::key (const unsigned int s) const
...
 
+unsigned int Tri6::which_node_am_i(unsigned int side,
+                                   unsigned int side_node) const
+{
+  libmesh_assert_less (side, this->n_sides());
+  libmesh_assert_less (side_node, 3);
+
+  return Tri6::side_nodes_map[side][side_node];
+}
+
```
Comment (roystgnr) [inline src/geom/face_tri6.C]: The dozens of copies of essentially-identical code here make me sad inside.  I can't think of an obvious way to improve it, though.  Make a virtual function that returns side_nodes_map (and one that returns nodes_on_side(s), for the second assert), and then make Elem::which_node_am_i non-virtual?

Comment (jwpeterson) [inline src/geom/face_tri6.C]: Well, one thing is that not all the asserts are the same (they depend on the Elem type), so there's not a ton of duplicate code savings there. Second, I thought about adding a virtual function that just returned the respective `Foo::side_nodes_map[s][n]` but decided that was about the same number of lines of code as this solution...

File: tests/geom/which_node_am_i_test.C (hunk 2 above)
Comment (roystgnr) [inline tests/geom/which_node_am_i_test.C]: It'd be good to have an example of that trick.  You've got the right philosophy in mind; if we *really* want decent test coverage someday then we'll need to hit the "throws an error" code paths deliberately.

But leave this PR as is.  If that trick isn't easy, then I messed up with the stream shim design somehow and I should be the one to fix it.

Also, I need to figure out how to debug broken unit tests properly.  For applications I usually just set a breakpoint at MPI_Abort, but that doesn't work with unit tests that catch errors and continue.  For our unit tests I used to "catch throw", but ever since we added the convergence_failure catch with pyramids I have to comment those FE tests out to avoid false positives, which is obviously a hack that doesn't scale.

**Comment by roystgnr** [comment]
I'm done nitpicking.  Fix as much or as little of that as you like; this looks good to merge as-is.

**Comment by jwpeterson** [comment]
> I'm done nitpicking. Fix as much or as little of that as you like; this looks good to merge as-is.

Thanks for reviewing! I'm pushing some of the simpler fixes in a moment.

File: tests/geom/which_node_am_i_test.C (hunk 2 above)
Comment (pbauman) [inline tests/geom/which_node_am_i_test.C]: > Also, I need to figure out how to debug broken unit tests properly. For applications I usually just set a breakpoint at MPI_Abort, but that doesn't work with unit tests that catch errors and continue. For our unit tests I used to "catch throw", but ever since we added the convergence_failure catch with pyramids I have to comment those FE tests out to avoid false positives, which is obviously a hack that doesn't scale.

I know there was some discussion of Google Test the other day, but I came across [Catch](https://github.com/philsquared/Catch) the other day (I think I saw @dmcdougall starred it, but not sure if he's actually tried it), but one of the neat looking things about it is that it looks like you can defines "modules" within the test suite and then only run those modules. I could see that being helpful on this point, e.g. only running `DistributedMesh` tests, or some such. But maybe Google Test has a similar feature?

Comment (jwpeterson) [inline tests/geom/which_node_am_i_test.C]: > But maybe Google Test has a similar feature?

Yes, each test is declared to be part of a particular "group", e.g.
```
TEST(LinearInterpolationTest, getSampleSize) {...}
TEST(LinearInterpolationTest, sample) {...}
```
and then you can run the GoogleTest executable with `--gtest_filter=LinearInterpolationTest` to run all tests from a given group.

We've already switched MOOSE over to using GoogleTest and it's been working well so far, hence my preference for that approach.

Comment (pbauman) [inline tests/geom/which_node_am_i_test.C]: That's great! How bad was the test refactoring for the switch? e.g. how hard will it be to switch libMesh?

Comment (jwpeterson) [inline tests/geom/which_node_am_i_test.C]: Converting the tests is very straightforward but it is a bit time consuming. One of the nice things about GoogleTest is that not everything has to be class-based (aka test fixtures) like cppunit, you can just declare free functions using the syntax I mentioned above for tests that don't require any state variables, so the tests often get shorter as well.

In MOOSE, we ended up with about half as many lines of code (idaholab/moose@bcec359) after the conversion.
